### PR TITLE
Fix hasher

### PR DESCRIPTION
**Pull request by marioortizmanero**
This PR finally fixes what I was experiencing in #83.

Turns out that the `HasherObject` implementation assumes that, in `Hasher`,  all the `write_TYPE` functions delegate their implementation to `write`. As in, it is only necessary to have an FFI-safe callback for `write`. This is true [in the standard library](https://doc.rust-lang.org/std/collections/hash_map/struct.DefaultHasher.html#impl-Hasher), where `write_u8` and similars rely on the default implementation of the `Hasher` trait. However, as soon as the hasher is changed to one that doesn't do this, `HasherObject` stops working as intended.

The [`AHash` hasher](https://docs.rs/ahash/latest/ahash/struct.AHasher.html#impl-Hasher) overrides `write_TYPE` so that, for example, `write_u8` delegates to `write_u64` instead of `write` for performance reasons. If someone hashes a `u8` with `Hasher` and then with `HasherObject`, they will be different. `Hasher` will call `write_u64`, but `HasherObject` will call `write`.

This issue propagated to other structs like `MapKey` or `MapQuery`, breaking more parts of `abi_stable`. It can be fixed by just implementing each `write_TYPE` function in `HasherObject`, rather than just `write`.

The only part that's left is implementing the `write_{i,u}128` functions, for which I would love to have your opinion, @rodrimati1992. Should I implement them as opaque types somehow (not sure if that would be possible)? Or do I throw a `panic!` in there indicating that it's unsupported?

Thanks to @darach, @Licenser, and @mfelsche for their help in solving this issue!

### CODE REVIEW

File: abi_stable/src/erased_types/trait_objects.rs [hunk 1]
```diff
@@ -21,7 +21,21 @@ use crate::{
 #[derive(StableAbi)]
 pub struct HasherObject<'a> {
     this: RMut<'a, ErasedObject>,
-    hash_slice: unsafe extern "C" fn(RMut<'_, ErasedObject>, RSlice<'_, u8>),
+    write: unsafe extern "C" fn(RMut<'_, ErasedObject>, RSlice<'_, u8>),
```
Comment (rodrimati1992) [inline abi_stable/src/erased_types/trait_objects.rs:24]: These write fields should move to a `&'static WriteFns` field with how many there are.

My general idea for the `WriteFns` struct
```rust
struct WriteFns{
    write: unsafe extern "C" fn(RMut<'_, ErasedObject>, RSlice<'_, u8>),
    write_i16: unsafe extern "C" fn(RMut<'_, ErasedObject>, i16),
    .... and all the other write_* function pointers
}

struct MakeWriteFns<T>;

impl<T> MakeWriteFns<T> 
where
    T: Hasher
{
    const V: &'static WriteFns = &WriteFns {
        write: write_Hasher::<T>,
        ....
    };
}
```

Comment (marioortizmanero) [reply abi_stable/src/erased_types/trait_objects.rs:24]: Yeah, that definitely looks tidier. Let me know if the last commit is what you expected.

**Comment by marioortizmanero** [comment]
@rodrimati1992, would it be possible to merge this & release v0.11? Or what's the schedule for that release? Do you have anything else planned?

File: abi_stable/src/erased_types/trait_objects.rs [hunk 2]
```diff
@@ -45,21 +48,96 @@ impl<'a> HasherObject<'a> {
...
+        (isize, write_isize),
+        (u16, write_u16),
+        (u32, write_u32),
+        (u64, write_u64),
+        (u8, write_u8),
+        (usize, write_usize)
+    );
+
+    fn write_i128(&mut self, val: i128) {
+        todo!()
```
Comment (rodrimati1992) [inline abi_stable/src/erased_types/trait_objects.rs:89]: Why `todo!()`, doesn't the default implementation work here(and for `write_u128`)?

Comment (marioortizmanero) [reply abi_stable/src/erased_types/trait_objects.rs:89]: It's in the other file, but commented out. As far as I know, there's no support for `i128` or `u128`, so these just don't compile when you uncomment it:

```rust
error[E0599]: the associated item `STABLE_ABI` exists for struct `GetTypeLayoutCtor<i128>`, but its trait bounds were not satisfied
   --> abi_stable/src/internal_macros.rs:6:37
    |
6   |         __GetTypeLayoutCtor::<$ty>::STABLE_ABI
    |                                     ^^^^^^^^^^ associated item cannot be called on `GetTypeLayoutCtor<i128>` due to unsatisfied trait bounds
    |
   ::: abi_stable/src/abi_stability/stable_abi_trait.rs:136:1
    |
136 | pub struct GetTypeLayoutCtor<T>(T);
    | ----------------------------------- associated item `STABLE_ABI` not found for this
    |
   ::: abi_stable/src/erased_types/trait_objects.rs:104:10
    |
104 | #[derive(StableAbi)]
    |          --------- in this derive macro expansion
    |
    = note: the following trait bounds were not satisfied:
            `i128: stable_abi_trait::StableAbi`
    = note: this error originates in the macro `$crate::_sabi_type_layouts` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0599]: the associated item `STABLE_ABI` exists for struct `GetTypeLayoutCtor<u128>`, but its trait bounds were not satisfied
   --> abi_stable/src/internal_macros.rs:6:37
    |
6   |         __GetTypeLayoutCtor::<$ty>::STABLE_ABI
    |                                     ^^^^^^^^^^ associated item cannot be called on `GetTypeLayoutCtor<u128>` due to unsatisfied trait bounds
    |
   ::: abi_stable/src/abi_stability/stable_abi_trait.rs:136:1
    |
136 | pub struct GetTypeLayoutCtor<T>(T);
    | ----------------------------------- associated item `STABLE_ABI` not found for this
    |
   ::: abi_stable/src/erased_types/trait_objects.rs:104:10
    |
104 | #[derive(StableAbi)]
    |          --------- in this derive macro expansion
    |
    = note: the following trait bounds were not satisfied:
            `u128: stable_abi_trait::StableAbi`
    = note: this error originates in the macro `$crate::_sabi_type_layouts` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0599`.
warning: `abi_stable` (lib) generated 2 warnings
error: could not compile `abi_stable` due to 2 previous errors; 2 warnings emitted
```

I did this a while ago so I might have forgotten what was fully going on, though. But it seems like these types don't implement `StableAbi`, so they can't be in a function in `WriteFns`.

**Pull request merged**